cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -555,6 +556,94 @@ public void programSpecification(HttpRequest request, HttpResponder responder,
   }
 
   /**
+   * Update the log level for a program according to the request body. Currently supported program types are
+   * {@link ProgramType#FLOW}, {@link ProgramType#SERVICE} and {@link ProgramType#WORKER}.
+   * The request body is expected to contain a map of log levels, where key is loggername, value is one of the
+   * valid {@link org.apache.twill.api.logging.LogEntry.Level} or null.
+   */
+  @Path("/apps/{app-name}/{program-type}/{program-name}/logLevels")
+  @PUT
+  public void updateProgramLogLevels(HttpRequest request, HttpResponder responder,
+                                     @PathParam("namespace-id") String namespace,
+                                     @PathParam("app-name") String appName,
+                                     @PathParam("program-type") String type,
+                                     @PathParam("program-name") String programName) throws Exception {
+    updateProgramLogLevels(request, responder, namespace, appName, ApplicationId.DEFAULT_VERSION, type, programName);
+  }
+
+  /**
+   * Update the log level for a program according to the request body.
+   */
+  @Path("/apps/{app-name}/versions/{app-version}/{program-type}/{program-name}/logLevels")
+  @PUT
+  public void updateProgramLogLevels(HttpRequest request, HttpResponder responder,
+                                     @PathParam("namespace-id") String namespace,
+                                     @PathParam("app-name") String appName,
+                                     @PathParam("app-version") String appVersion,
+                                     @PathParam("program-type") String type,
+                                     @PathParam("program-name") String programName) throws Exception {
+    ProgramType programType = getProgramType(type);
+    if (programType == null) {
+      throw new MethodNotAllowedException(request.getMethod(), request.getUri());
+    }
+    try {
+      // we are decoding the body to Map<String, String> instead of Map<String, LogEntry.Level> here since Gson will
+      // serialize invalid enum values to null, which is allowed for log level, instead of throw an Exception.
```

Comment:
Handlers are the places where we should sanitize all user provided input. You should validate that the log levels provided here are valid log levels and throw a BadRequestException if not. The lifecycleService should not have to perform any validation, and should receive Map<String, LogEntry.Level> instead of a possibly invalid map.